README.md
# Atlassian Confluence for Docker

__docker-atlassian-confluence__ provides [Atlassian Confluence](https://www.atlassian.com/software/confluence) in a [docker](https://www.docker.com/)
container to support team collaboration.

## Related Projects

You may also be interested in the following projects:

  * [Dockerized Atlassian Crowd](https://github.com/markeissler/docker-atlassian-crowd) - centralized identity management
  * [Dockerized Atlassian JIRA](https://github.com/markeissler/docker-atlassian-jira) - project tracking for software teams


## Overview

This revision of __docker-atlassian-confluence__ will install:

[Confluence 6.5.0](https://confluence.atlassian.com/doc/issues-resolved-in-6-5-0-939699975.html)

## Installation

This application is ready to launch on a Docker host:

```sh
prompt> docker run -d -p 8080:8080 -p 8443:8443 markeissler/atlassian-confluence:latest
```

## Usage

<a name="data-persistence"></a>

### Data Persistence

As configured, data on the following volumes will be created to persist data between container starts:

| Volume | Purpose                                                    |
|:-------|:-----------------------------------------------------------|
| /var/atlassian/confluence               | application configuration |
| /opt/atlassian/confluence/logs          | runtime logs              |

<a name="data-persistence-nfs"></a>

### Data Persistence over NFS

It may be desirable to configure data persistence over NFS, in which case NFS volumes are mounted at the locations
described in the [Data Persistence](#data-persistence) section above. NFS support requires that the underlying Docker
host supports NFS; if deploying to a [Docker swarm](https://docs.docker.com/engine/swarm/) a potential __boot2docker.iso__
candidate that supports NFS is the [boot2docker-nfs.iso](https://github.com/markeissler/boot2docker-nfs).

Certain JIRA directories are moved out of the application configuration directory and into an ephemeral runtime storage
area to prevent data corruption startup failures. Specfically, cache directories are moved so that clean re-starts
are possible; often, when an instance dies Tomcat will not be shutdown cleanly and data corruption is likely to occur
with regard to the _felix_ plugin cache).

| Directory | Purpose                                                        |
|:----------|:---------------------------------------------------------------|
| /var/atlassian/confluence_runtime | runtime storage for caches and indexes |

### SSL Support

You can enable SSL by simply copying a PKCS12 format certificate (`certificate.p12`) into the `CONF_HOME` directory
(`/var/atlassian/confluence`) and then restarting the container. The PKCS12 file format has been selected to make it
easier to generate certificates using `openssl`.

An example `openssl` command that will create a PKCS12 file from a private key (`server_key.pem`) and public certficate
(`server_cert.pem`) follows:

```sh
prompt> openssl pkcs12 -export -in server_cert.pem \
    -inkey server_key.pem -out certificate.p12 \
    -passout pass:changeit -name "confluence"
```

On container startup, the PCKS12 format certificate.p12 file will be converted and stored in the system JKS keystore.

## Docker Swarm Support

While __docker-atlassian-confluence__ does not support multi-node clustering it does support deployment to a cluster
with a failover configuration (where only a single Confluence instance is active at any time).

This configuration requires that [Data Persistence over NFS](#data-persistence-nfs) has been configured to share
Confluence configuration information among replicated instances.

## Redeployment Index Regeneration

Indexes are stored in ephemeral storage; consequently, when redeploying the application on top of existing persistent
storage (i.e. [Data Persistence over NFS](#data-persistence-nfs)) it will be necessary to manually trigger a rebuild of
all indexes. Log in as a user with administrative permissions, then navigate to:

`Cog > General Configuration > Administration > Content Indexing`

Click on the <kbd>rebuild</kbd> button.

## Troubleshooting

For general troubleshooting information check the [Troubleshoot](troubleshoot.md) document.

## Upstream Release Notes

Release notes for all versions of Confluence can be viewed online:

[Confluence Release Notes](https://confluence.atlassian.com/doc/confluence-release-notes-327.html)

## Authors

__docker-atlassian-confluence__ is the work of __Mark Eissler__.

## Attributions

__docker-atlassian-confluence__ was inspired by the work of [Martin Aksel Jensen](https://github.com/cptactionhank),
specifically his ongoing efforts to provide up-to-date _dockerized_ versions of other popular [Atlassian](https://www.atlassian.com/)
applications.

## License

__docker-atlassian-confluence__ is licensed under the MIT open source license.

---
Without open source, there would be no Internet as we know it today.


Changelog.md
# Changelog: docker-atlassian-confluence

## 1.11.0 / 2018-01-28

Update Confluence to 6.5.0.

### Minor version increment!

Since this is a Minor version increment be sure to review the release notes and upgrade notes:

  * [Confluence 6.5 upgrade notes] (https://confluence.atlassian.com/doc/confluence-6-5-upgrade-notes-938862633.html)
  * [Confluence 6.5 release notes] (https://confluence.atlassian.com/doc/confluence-6-5-release-notes-938862614.html)
  * [Confluence 6.5.0 issues resolved] (https://confluence.atlassian.com/doc/issues-resolved-in-6-5-0-939699975.html)

## 1.10.0 / 2017-11-05

Update Confluence to 6.4.3.

## 1.9.0 / 2017-11-05

Update Confluence to 6.4.2.

## 1.8.0 / 2017-11-05

Update Confluence to 6.4.1.

## 1.7.0 / 2017-09-14

Update Confluence to 6.4.0.

### Minor version increment!

Since this is a Minor version increment be sure to review the release notes and upgrade notes:

  * [Confluence 6.4 upgrade notes] (https://confluence.atlassian.com/doc/confluence-6-4-upgrade-notes-934721251.html)
  * [Confluence 6.4 release notes] (https://confluence.atlassian.com/doc/confluence-6-4-release-notes-934721238.html)
  * [Confluence 6.4.0 issues resolved] (https://confluence.atlassian.com/doc/issues-resolved-in-6-4-0-937165586.html)

> NOTE: Breaking tradition with Atlassian Minor version updates, this time the team did not skip the `.0` release.

## 1.6.0 / 2017-09-14

Update Confluence to 6.3.4.

## 1.5.0 / 2017-09-13

Update Confluence to 6.3.3.

## 1.4.0 / 2017-09-13

Update Confluence to 6.3.2.

## 1.3.0 / 2017-07-30

Update Confluence to 6.3.1.

### Minor version increment!

Since this is a Minor version increment be sure to review the release notes and upgrade notes:

  * [Confluence 6.3 upgrade notes] (https://confluence.atlassian.com/doc/confluence-6-3-upgrade-notes-909642703.html)
  * [Confluence 6.3 release notes] (https://confluence.atlassian.com/doc/confluence-6-3-release-notes-909642701.html)
  * [Confluence 6.3.1 issues resolved] (https://confluence.atlassian.com/doc/issues-resolved-in-6-3-1-931235734.html)

> NOTE: As is common with Atlassian Minor version updates, there is no 6.3.0 release.

### MySQL driver updated to 5.1.42

The included **MySQL** driver has been updated to `5.1.42` in conjunction with included support in this release:

[Updated database drivers](https://confluence.atlassian.com/doc/confluence-6-3-upgrade-notes-909642703.html#Confluence6.3UpgradeNotes-Updateddatabasedrivers)

Atlassian bundles the **Postgres** driver with each release. This release includes version `42.1.1` of the **Postgres** driver.

### Short list of commit messages

  * Update mysql driver to 5.1.42.
  * Update Confluence to 6.3.1.

## 1.2.0 / 2017-07-28

Update Confluence to 6.2.4.

### Short list of commit messages.

  * Update Confluence to 6.2.4.

## 1.1.0 / 2017-06-24

Update Confluence to 6.2.3.

### Short list of commit messages

  * Update README to include link to upstream release notes.
  * Update Confluence to 6.2.3.

## 1.0.0 / 2017-06-24

Update Confluence to 6.2.2.

### Short list of commit messages

  * Update README for redeployment indexing issues.
  * Update Confluence to 6.2.2.

## 0.10.0 / 2017-06-20

Docker Swarm support! This version adds support for deployment to a cluster with a failover configuration. That is, only
one instance can be active at a time but the failover instance should startup without encountering errors stemming from
a corrupted _felix_ plugin cache.

Indexes have been moved out of the Docker host's persistence directory to improve performance in configurations that
enable data persistence over NFS volumes.

### Short list of commit messages

  * Update README for ephemeral storage and Swarm support.
  * Use ephemeral storage for caches
  * Update README for SSL support.

## 0.9.1 / 2017-06-18

Maintenance release.

### Short list of commit messages

  * Update README for v0.9.1. Fix app description.

## 0.9.0 / 2017-06-16

Initial release! A _dockerized_ [Atlassian Confluence](https://www.atlassian.com/software/confluence) install.

### Short list of commit messages

  * Update README for v0.9.0.


docker-entrypoint.sh
#!/bin/bash

# check if the `server.xml` file has been changed since the creation of this
# Docker image. If the file has been changed the entrypoint script will not
# perform modifications to the configuration file.
if [ "$(stat --format "%Y" "${CONF_INSTALL}/conf/server.xml")" -eq "0" ]; then
  if [ -n "${X_PROXY_NAME}" ]; then
    xmlstarlet ed --inplace --pf --ps --insert '//Connector[@port="8080"]' --type "attr" --name "proxyName" --value "${X_PROXY_NAME}" "${CONF_INSTALL}/conf/server.xml"
  fi
  if [ -n "${X_PROXY_PORT}" ]; then
    xmlstarlet ed --inplace --pf --ps --insert '//Connector[@port="8080"]' --type "attr" --name "proxyPort" --value "${X_PROXY_PORT}" "${CONF_INSTALL}/conf/server.xml"
  fi
  if [ -n "${X_PROXY_SCHEME}" ]; then
    xmlstarlet ed --inplace --pf --ps --insert '//Connector[@port="8080"]' --type "attr" --name "scheme" --value "${X_PROXY_SCHEME}" "${CONF_INSTALL}/conf/server.xml"
  fi
  if [ -n "${X_PROXY_SECURE}" ]; then
    xmlstarlet ed --inplace --pf --ps --insert '//Connector[@port="8080"]' --type "attr" --name "secure" --value "${X_PROXY_SECURE}" "${CONF_INSTALL}/conf/server.xml"
  fi
  if [ -n "${X_PATH}" ]; then
    xmlstarlet ed --inplace --pf --ps --update '//Context[@docBase="../confluence"]/@path' --value "${X_PATH}" "${CONF_INSTALL}/conf/server.xml"
  fi
fi

if [ -f "${CERTIFICATE}" ] || [ -f "${CERTIFICATE}.p12" ]; then
  # convert PKCS12 certificate format to JKS certificate format
  #
  # To generate a pkcs12 file from an openssl self-signed cert and key file:
  #   > openssl pkcs12 -export -in server_cert.pem -inkey server_key.pem -out certificate.p12
  #       -passout pass:changeit -name "confluence"
  #
  # To test the insertion:
  #   > docker exec -it <CONTAINER_ID> /bin/bash
  #   > keytool -list -keystore $JAVA_HOME/jre/lib/security/cacerts -v | grep Alias | grep confluence
  #
  if [[ "${CERTIFICATE}" =~ .p12$ || -f "${CERTIFICATE}.p12" ]]; then
    keytool -noprompt -storepass changeit -importkeystore \
      -srckeystore ${CERTIFICATE%.p12}.p12 -srcstoretype PKCS12 -srcstorepass changeit -alias confluence \
      -destkeystore ${JAVA_CACERTS} -deststoretype JKS -deststorepass changeit -destalias confluence
  else
    keytool -noprompt -storepass changeit \
      -keystore ${JAVA_CACERTS} -import -file ${CERTIFICATE} -alias confluence
  fi

  # Update the server.xml file
  # <!--
  # <Connector port="8443" maxHttpHeaderSize="8192"
  #    maxThreads="150" minSpareThreads="25"
  #    protocol="org.apache.coyote.http11.Http11NioProtocol"
  #    enableLookups="false" disableUploadTimeout="true"
  #    acceptCount="100" scheme="https" secure="true"
  #    clientAuth="false" sslProtocol="TLSv1.2" sslEnabledProtocols="TLSv1.2" SSLEnabled="true"
  #    URIEncoding="UTF-8" keystorePass="<MY_CERTIFICATE_PASSWORD>"/>
  # -->
  xmlstarlet ed --inplace --pf --ps \
    --subnode "Server/Service" --type elem --name "ConnectorTMP" --value "" \
    --insert  "//ConnectorTMP" --type attr --name "port" --value "8443" \
    --insert  "//ConnectorTMP" --type attr --name "protocol" --value "org.apache.coyote.http11.Http11NioProtocol" \
    --insert  "//ConnectorTMP" --type attr --name "maxHttpHeaderSize" --value "8192" \
    --insert  "//ConnectorTMP" --type attr --name "SSLEnabled" --value "true"  \
    --insert  "//ConnectorTMP" --type attr --name "maxThreads" --value "150" \
    --insert  "//ConnectorTMP" --type attr --name "minSpareThreads" --value "25" \
    --insert  "//ConnectorTMP" --type attr --name "enableLookups" --value "false" \
    --insert  "//ConnectorTMP" --type attr --name "disableUploadTimeout" --value "true" \
    --insert  "//ConnectorTMP" --type attr --name "acceptCount" --value "100" \
    --insert  "//ConnectorTMP" --type attr --name "scheme" --value "https" \
    --insert  "//ConnectorTMP" --type attr --name "secure" --value "true" \
    --insert  "//ConnectorTMP" --type attr --name "clientAuth" --value "false" \
    --insert  "//ConnectorTMP" --type attr --name "sslProtocol" --value "TLSv1.2" \
    --insert  "//ConnectorTMP" --type attr --name "sslEnabledProtocols" --value "TLSv1.2" \
    --insert  "//ConnectorTMP" --type attr --name "useBodyEncodingForURI" --value "true" \
    --insert  "//ConnectorTMP" --type attr --name "keyAlias" --value "confluence" \
    --insert  "//ConnectorTMP" --type attr --name "keystoreFile" --value "${JAVA_CACERTS}" \
    --insert  "//ConnectorTMP" --type attr --name "keystorePass" --value "changeit" \
    --insert  "//ConnectorTMP" --type attr --name "keystoreType" --value "JKS" \
    --rename  "//ConnectorTMP" --value "Connector" \
    "${CONF_INSTALL}/conf/server.xml"

  # @TODO: Update Base URL to HTTPS
  #
  # https://confluence.atlassian.com/confkb/how-do-i-manually-change-the-base-url-310378877.html
  #

  # @TODO: Use xmlstarlet to update the web.xml file
  #
  # This will redirect all traffic to use HTTPS urls.
  #
  # Location: ${CONF_INSTALL}/confluence/WEB-INF/web.xml
  # <!--
  # <security-constraint>
  #   <web-resource-collection>
  #     <web-resource-name>Restricted URLs</web-resource-name>
  #     <url-pattern>/</url-pattern>
  #   </web-resource-collection>
  #   <user-data-constraint>
  #     <transport-guarantee>CONFIDENTIAL</transport-guarantee>
  #   </user-data-constraint>
  # </security-constraint>
  # -->
fi

exec "$@"
